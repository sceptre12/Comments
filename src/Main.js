import React, { useState, useEffect, useRef } from "react";
import moment from "moment";

// Components
import CommentsFeed from "./components/commentsFeed";
import CommentCreator from "./components/commentCreator";

// Api
import { getComments, getComment } from "./api/endPoints";
// Style
import "./main.css";

const defaultCommentSubmissionState = {
  wasSubmitted: false,
  commentId: -1,
};
const defaultPostCommentState = {
  hasRetrievedNewComment: false, // Gets Set whenever a posted Comment was retrieved from the db
  commentId: -1,
};

export const sortDateDesc = (commentA, commentB) => {
  const isSame = moment(commentA.created).isSame(commentB.created);
  const isAfter = moment(commentA.created).isAfter(commentB.created);

  if (isSame) {
    return 0;
  } else {
    if (isAfter) {
      return -1;
    } else {
      return 1;
    }
  }
};

const Main = () => {
  // Main Comments state
  const [commentsState, setCommentsState] = useState({
    comments: [],
    isLoadingEverything: true,
  });
  // State for determining a user has submitted a post
  const [commentSubmissionState, setIfCommentWasSubmitted] = useState({
    ...defaultCommentSubmissionState,
  });
  // Posted Comment State
  const [postedCommentState, setPostedCommentState] = useState({
    ...defaultPostCommentState,
  });
  // Style Theme check
  const [isInDarkMode, setIfInDarkMode] = useState(false);

  /*
    This is used to store the timeoutID for resetting 
    the postedCommentState
  */
  const resetTimer = useRef(undefined);

  // Runs on first mount
  useEffect(() => {
    getComments()
      .then((result) => {
        result.sort(sortDateDesc);
        setCommentsState({
          comments: result,
          isLoadingEverything: false,
        });
      })
      .catch((err) => console.error(err));
  }, []);

  /**
   * Checks if a post was submitted
   * If so then retrieve that post from the db
   */
  useEffect(() => {
    if (commentSubmissionState.wasSubmitted) {
      getComment(commentSubmissionState.commentId)
        .then((result) => {
          if (!!resetTimer.current) {
            clearTimeout(resetTimer.current);
            resetTimer.current = undefined;
          }

          // Reset Submission state
          setIfCommentWasSubmitted({ ...defaultCommentSubmissionState });

          // Update the comment state
          setCommentsState({
            ...commentsState,
            comments: [{ ...result }, ...commentsState.comments],
          });

          // Set Posted Comment State
          setPostedCommentState({
            hasRetrievedNewComment: true,
            commentId: commentSubmissionState.commentId,
          });
        })
        .catch((err) => console.error(err));
    }
  }, [commentSubmissionState, commentsState]);

  // Resets Posted Comment State
  useEffect(() => {
    if (postedCommentState.hasRetrievedNewComment) {
      resetTimer.current = setTimeout(() => {
        setPostedCommentState({
          ...defaultPostCommentState,
        });
      }, 2000);
    }
  }, [postedCommentState.hasRetrievedNewComment]);

  return (
    <div id="main">
      <CommentCreator {...{ setIfInDarkMode, setIfCommentWasSubmitted }} />
      <CommentsFeed
        {...{
          isInDarkMode,
          ...commentsState,
          postedCommentState,
        }}
      />
    </div>
  );
};

export default Main;
