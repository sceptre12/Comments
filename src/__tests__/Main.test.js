import React from "react";
import { shallow } from "enzyme";
import moment from "moment";
import Main, { sortDateDesc } from "../Main";

describe("Testing Main.js", () => {
  const container = shallow(<Main />);
  it("Testing Main.js Snapshot", () => {
    expect(container.html()).toMatchSnapshot();
  });

  it("Testing Date Sorting", () => {
    let timeStamp1 = moment().subtract(56, "hours");
    let timeStamp2 = moment().subtract(20, "hours");
    const tempDateList = [
      { created: timeStamp1.toDate() },
      { created: timeStamp2.toDate() },
    ];
    tempDateList.sort(sortDateDesc);
    expect(tempDateList).toEqual([
      { created: timeStamp2.toDate() },
      { created: timeStamp1.toDate() },
    ]);
  });
});
