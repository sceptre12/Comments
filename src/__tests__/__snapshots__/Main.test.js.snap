// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Main.js Testing Main.js Snapshot 1`] = `"<div id=\\"main\\"><aside id=\\"comment-creator\\"><div id=\\"comment-form-container\\"><div id=\\"comment-form_header\\"><h2>Comments Feed</h2></div><form id=\\"comment-form\\"><div id=\\"comment-form-fieldset\\"><div id=\\"comment-form-name\\"><div class=\\"comment-form-label\\"><label for=\\"comment-name\\">Name:</label></div><div class=\\"comment-form-userinput\\"><input type=\\"text\\" id=\\"comment-name\\" value=\\"\\"/></div></div><div id=\\"comment-form-message\\"><div class=\\"comment-form-label\\"><label for=\\"comment-message\\">Message:</label></div><div class=\\"comment-form-userinput\\"><textarea name=\\"comment-message\\" id=\\"comment-message\\" maxLength=\\"160\\"></textarea></div></div></div><div id=\\"comment-send-button_container\\"><button type=\\"submit\\" disabled=\\"\\"><span>Send</span></button></div></form></div></aside><main id=\\"comments-feed\\"><div id=\\"comments-feed_loadingSpinner\\"><div class=\\"css-9ebb9a\\"><div class=\\"css-119vicx\\"></div><div class=\\"css-7qd4qk\\"></div><div class=\\"css-1ol73ma\\"></div><div class=\\"css-1ch7kc6\\"></div><div class=\\"css-1bs6bws\\"></div><div class=\\"css-jq7hsu\\"></div></div></div></main></div>"`;
